opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -2485,143 +2485,258 @@ bool CpGrid::adapt(const std::vector<std::array<int,3>>& cells_per_dim_vec,
+    const auto& globalIdSet = this->globalIdSet();
 
-    return preAdapt();
+    parentToFirstChildGlobalIds.resize(data[0]->size(0), -1); // Initialize with -1 (invalid value, for non parent cells).
+
+    const auto& elements = Dune::elements(levelGridView(0));
+    for (const auto& element : elements) {
+        const auto& [level, children] = parentToChildrenBeforeLoadBalance[element.index()];
+
+        if (!children.empty()) {
+            const auto& levelData = *data[level];
+            const auto& first_child = Dune::cpgrid::Entity<0>(levelData, children[0], true);
+
+            // Rewrite parent global id entry with first child global id
+            parentToFirstChildGlobalIds[globalIdSet.id(element)] = globalIdSet.id(first_child);
+        }
+    }
+#endif
 }
 
-void CpGrid::postAdapt()
+void CpGrid::syncCellIds()
 {
-    // - Resize with the new amount of cells on the leaf grid view
-    // - Set marks equal to zero (representing 'doing nothing')
-    current_view_data_ -> postAdapt();
+#if HAVE_MPI
+    std::vector<int> parentToFirstChildGlobalIds;
+    getFirstChildGlobalIds(parentToFirstChildGlobalIds);
+
+    auto size = comm().max(parentToFirstChildGlobalIds.size());
+
+    switchToDistributedView();
+
+    std::vector<int> firstChildGlobalIds(size);
+    if (comm().rank() == 0) {
+        firstChildGlobalIds = parentToFirstChildGlobalIds;
+    }
+    comm().broadcast(firstChildGlobalIds.data(), firstChildGlobalIds.size(), 0);
+
+    const int maxLevel = this->maxLevel();
+
+    // Preallocate syncCellIds
+    std::vector<std::vector<int>> syncCellIds(maxLevel);
+    for (int level = 1; level <= maxLevel; ++level) {
+        syncCellIds[level-1].resize(currentData()[level]->size(0));
+    }
+
+    const auto& globalIdSet = this->globalIdSet();
+
+    // Populate for interior cells
+    for (int level = 1; level <= maxLevel; ++level) {
+        const auto& elements = Dune::elements(levelGridView(level), Dune::Partitions::interior);
+        for (const auto& element : elements) {
+            const int parent_globalId = globalIdSet.id(element.father());
+            const int idx_in_parent = element.getIdxInParentCell();
+            const int first_child_id = firstChildGlobalIds[parent_globalId];
+            const int new_elem_globalId = first_child_id + idx_in_parent;
+
+            syncCellIds[element.level()-1][element.index()] = new_elem_globalId;
+        }
+    }
+
+    // Syncronize ids via data Handle
+    const auto& parent_to_children = currentData().front()->getParentToChildren();
+    ParentToChildrenCellGlobalIdHandle parentToChildrenCellGlobalId_handle(parent_to_children,
+                                                                           syncCellIds);
+    currentData().front()->communicate(parentToChildrenCellGlobalId_handle,
+                                       Dune::InteriorBorder_All_Interface,
+                                       Dune::ForwardCommunication );
```

Comment:
This would not be needed if iterate over all elements in the loop above (and not only interior). I think we should do it like this.

Reply:
Nice change!